### ISSUE TITLE
Bug: Potential bug in “official” useInterval example

### ISSUE THREAD

**Issue by gmoniava** [facebook/react#21912]
I know this maybe not react related but it is related to a blog post by react team member.
The bug is explained here: https://stackoverflow.com/questions/68407187/potential-bug-in-official-useinterval-example
I think feedback from react team would be useful.

**Comment by bvaughn** [comment]
The report on Stack Overflow doesn't look like a bug to me.

The answer that suggests updating the ref during render would be a side effect, which [should be avoided because it will cause problems](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects).

> The demo shows that the current callback value may differ from that in useEffect alright, but the real question is which one of them is the "correct" one?

The "correct" one is the one that has been committed. For one reason, committed effects are the only ones that are guaranteed to have cleanup phase later. (The interval in this question doesn't _need_ a cleanup effect, but other things might.)

Another more compelling reason in this case, perhaps, is that React may pre-render things (either at lower priority, or because they're "offscreen" and not yet visible, or in the future b'c of animation APIs). Pre-rendering work like this should never modify a ref, because the modification would be arbitrary. (Consider a future animation API that pre-renders multiple possible future visual states to make transitions faster in response to user interaction. You wouldn't want the one that happened to render last to just mutate a ref that's used by your currently visible/committed view.)

**Comment by gmoniava** [comment]
@bvaughn AFAIK the answer doesn't suggest to update the ref during render, the author was just contemplating; he states that he knows it is a bad idea to update it in render because of async rendering (bottom of question). Regardless, the custom `useInterval` still seems to be buggy.

**Comment by bvaughn** [comment]
I think that thinking of it as an "old" callback is maybe the wrong mental model. At some point in the future, after an application updates, any callback is an "old" callback.

**Comment by gmoniava** [comment]
@bvaughn FYI another developer similarly reported similar issue as a bug: https://stackoverflow.com/questions/61999810/store-a-callback-in-useref/68104782#68104782. I understand them, because AFAIK the goal of that approach Dan did in that code, was that, if someone passed a new callback to `useInterval` during render, then that callback was supposed to be executed instead of a one from previous render.

**Comment by bvaughn** [comment]
> if someone passed a new callback to `useInterval` during render, then that callback was supposed to be executed instead of a one from previous render.

The new callback prop should only be executed _if it was committed_.

**Comment by gmoniava** [comment]
@bvaughn Quote from Dan's blog:

> So what if we didn’t replace the interval at all, and instead introduced a mutable savedCallback variable pointing to the **latest** interval callback?

> However, thanks to the savedCallback ref, we can always read the callback that **we set after the last render**, and call it from the interval tick.

How about this?

**Comment by bvaughn** [comment]
How about what? I'm not sure I understand.

When Dan says, "_after the last render_", he literally means "after the render, in an effect" (as is shown in the code example right below that)

**Comment by gmoniava** [comment]
@bvaughn Ok, it seems confusion comes from the fact, which callback do we consider the _latest_.

I am also not sure with the terminology you used above _committed_, as I haven't started using React 17. Anyway, if there is something more to say I think we better do it on Stack Overflow thread, there is more visibility. 

What I am trying to say you can see there are at least three developers who got confused about this.

**Comment by gmoniava** [comment]
@bvaughn Ok, it seems confusion comes from the fact, which callback do we consider the latest. I am calling latest the one which was updated as a result of latest render (before useEffect was run).

I am also not very familiar with the concept you used above committed, as I haven't started using React 17.

What I am trying to say you can see there are at least three developers who got confused about this. And if someone was using that custom hook useInterval they might be expecting it to use the latest callback (latest in the sense as the one which they passed during last render), isn't it?

**Comment by bvaughn** [comment]
I think our conversation is kind of going in a circle.

Originally, you asked if this was a bug and I responded and said it wasn't, and tried to explain why. I also responded to the Stack Overflow as well.

I'm not sure what we're talking about now. It seems like you're concerned that the article is a little confusing, and maybe it is. Maybe it could have been clearer. That's not something I can address though.

I think I'm going to bow out of this conversation now 🙇🏼

**Comment by gmoniava** [comment]
@bvaughn Ok, my point is if I was a consumer of `useInterval` custom hook, I would expect it to run the latest callback - latest, in the sense of the one which I passed during last render, isn't it?

**Comment by bvaughn** [comment]
I think you'd have to take this up with Dan. I'm not sure what else to add. The way it's written seems intentional (no side effects during render).

**Comment by gmoniava** [comment]
@bvaughn I don't know how to reach him I don't have twitter, maybe you can show him the stack overflow question. At least you can see I am not the only one confused.

**Comment by gmoniava** [comment]
@bvaughn There is also a related question: https://stackoverflow.com/questions/61999810/store-a-callback-in-useref, in case you will be interested to write an answer for on SO. It doesn't have a official answer. And seems popular question.

**Comment by Izhaki** [comment]
I'm pretty sure that [the original `userInterval` code](https://overreacted.io/making-setinterval-declarative-with-react-hooks/) is buggy.

The following is ultimately a summary of all the contributions by others in [my SO question](https://stackoverflow.com/q/68407187/1179377).

# Render Phase Callbacks

```js
const [size, setSize] = React.useState();

const onInterval = () => {
  console.log(size)
}

useInterval(onInterval, 100);
```

`onInterval` in this example will only be the "correct" callback after the DOM has been committed. 

Concurrent mode has it that the render phase resulting in a new value for `onInterval` may not even make it to the DOM.

Calling `savedCallback.current = callback` outside an effect constitutes a side-effect (as @bvaughn pointed out), so such assignment must happen within an effect - as in the original `useInterval` implementation.

# A Fraction of Time

I think what @gaearon was missing is that the interval could fire *after* the DOM has been committed, but *before* `useEffect` sets `savedCallback.current = callback`.

Once the DOM has been committed, the incoming `callback` argument in `useInterval` is the one that should be used in the interval callback.

As @bvaughn noted. using `useLayoutEffect` instead of `useEffect` would solve this. So the `useInterval` code should change to:

```js
  // Remember the latest callback.
  useLayoutEffect(() => {
    savedCallback.current = callback;
  }, [callback]);
```

**Comment by Izhaki** [comment]
I reckon this issue should move to https://github.com/gaearon/overreacted.io/issues and be re-opened there.

**Comment by gmoniava** [comment]
@Izhaki I didn't know that repo existed, yeah it should be moved there. But since this issue is closed maybe it is better you open a new one there?

**Comment by gmoniava** [comment]
@Izhaki Yeah, it should go there. Will you open the new issue, or shall I do it?

**Comment by Izhaki** [comment]
Owners of this repo should have an option to "Move issue". Guess if it Ain't done in the next 48 hours we should just open a new one there. I don't mind doing it.

**Comment by bvaughn** [comment]
> Owners of this repo should have an option to "Move issue"

This option only works for other projects within the same org (facebook). The repo you mentioning above is not a part of the facebook org.
